graphbolt/src/csc_sampling_graph.cc

```diff
@@ -472,5 +498,95 @@ torch::Tensor PickByEtype(
   return torch::cat(picked_neighbors, 0);
 }
 
+template <>
+torch::Tensor Pick<sampler_t::LABOR>(
+    int64_t offset, int64_t num_neighbors, int64_t fanout, bool replace,
+    const torch::TensorOptions& options,
+    const torch::optional<torch::Tensor>& probs_or_mask,
+    sampler_args<sampler_t::LABOR> args) {
+  if (fanout == 0) return torch::tensor({}, options);
+  if (probs_or_mask.has_value()) {
+    torch::Tensor picked_neighbors;
+    AT_DISPATCH_FLOATING_TYPES(
+        probs_or_mask.value().scalar_type(), "LaborPickFloatType", ([&] {
+          picked_neighbors = LaborPick<true, scalar_t>(
+              offset, num_neighbors, fanout, args.random_seed, options,
+              probs_or_mask, args.indices);
+        }));
+    return picked_neighbors;
+  } else {
+    return LaborPick<false>(
+        offset, num_neighbors, fanout, args.random_seed, options, probs_or_mask,
+        args.indices);
+  }
+}
+
+template <bool nonuniform, typename fp_t = float>
+inline torch::Tensor LaborPick(
+    int64_t offset, int64_t num_neighbors, int64_t fanout, int64_t random_seed,
+    const torch::TensorOptions& options,
+    const torch::optional<torch::Tensor>& probs_or_mask,
+    const torch::Tensor& indices) {
+  fanout = fanout < 0 ? num_neighbors : std::min(fanout, num_neighbors);
+  if (!nonuniform && fanout >= num_neighbors) {
+    return torch::arange(offset, offset + num_neighbors, options);
+  }
+  torch::Tensor heap_tensor = torch::empty({fanout * 2}, torch::kInt32);
+  // Assuming max_degree of a vertex is <= 4 billion
+  auto heap_data = reinterpret_cast<std::pair<float, uint32_t>*>(
+      heap_tensor.data_ptr<int32_t>());
+  const fp_t* local_probs_data =
+      nonuniform ? probs_or_mask.value().data_ptr<fp_t>() + offset : nullptr;
+  AT_DISPATCH_INTEGRAL_TYPES(
+      indices.scalar_type(), "LaborPickMain", ([&] {
+        const scalar_t* local_indices_data =
+            indices.data_ptr<scalar_t>() + offset;
+        for (uint32_t i = 0; i < fanout; ++i) {
+          const auto t = local_indices_data[i];
+          pcg32 ng(random_seed, t);
+          std::uniform_real_distribution<float> uni;
+          auto rnd = uni(ng);  // r_t
```

Comment:
what does r_t mean?

Reply:
https://arxiv.org/pdf/2210.13339.pdf Page 13 algorithm 1 has r_t \leq c_s \pi_t. But here, I used Page 14, A3: Sampling a fixed number of neighbors.

Reply:
This code has heap functions etc. which are logarithmic but the complexity analysis yields a linear complexity :).